### Import Library

In [30]:
import numpy as np
import pandas as pd
from sklearn.ensemble import (  
    AdaBoostClassifier, RandomForestClassifier) # import RandomForest & Boost
from sklearn.metrics import accuracy_score, classification_report
from sklearn.model_selection import GridSearchCV, train_test_split
from sklearn.tree import DecisionTreeClassifier  # import DT

### Persiapan Data

In [31]:
# Load data
df = pd.read_csv('mushrooms.csv')

df.head()

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0     p         x           s         n       t    p               f   
1     e         x           s         y       t    a               f   
2     e         b           s         w       t    l               f   
3     p         x           y         w       t    p               f   
4     e         x           s         g       f    n               f   

  gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0            c         n          k  ...                        s   
1            c         b          k  ...                        s   
2            c         b          n  ...                        s   
3            c         n          n  ...                        s   
4            w         b          k  ...                        s   

  stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                      w                      w         p          w   
1                      w                      w         p          w   
2                      w                      w         p          w   
3                      w                      w         p          w   
4                      w                      w         p          w   

  ring-number ring-type spore-print-color population habitat  
0           o         p                 k          s       u  
1           o         p                 n          n       g  
2           o         p                 n          n       m  
3           o         p                 k          s       u  
4           o         e                 n          a       g  

[5 rows x 23 columns]

In [32]:
# Use LabelEncoder to make the columns into machine understandable format.
# Using LabelEncoder to convert catergory values to ordinal
from sklearn.preprocessing import LabelEncoder

labelencoder=LabelEncoder()
for column in df.columns:
    df[column] = labelencoder.fit_transform(df[column])

df.head()

class  cap-shape  cap-surface  cap-color  bruises  odor  gill-attachment  \
0      1          5            2          4        1     6                1   
1      0          5            2          9        1     0                1   
2      0          0            2          8        1     3                1   
3      1          5            3          8        1     6                1   
4      0          5            2          3        0     5                1   

   gill-spacing  gill-size  gill-color  ...  stalk-surface-below-ring  \
0             0          1           4  ...                         2   
1             0          0           4  ...                         2   
2             0          0           5  ...                         2   
3             0          1           5  ...                         2   
4             1          0           4  ...                         2   

   stalk-color-above-ring  stalk-color-below-ring  veil-type  veil-color  \
0                       7                       7          0           2   
1                       7                       7          0           2   
2                       7                       7          0           2   
3                       7                       7          0           2   
4                       7                       7          0           2   

   ring-number  ring-type  spore-print-color  population  habitat  
0            1          4                  2           3        5  
1            1          4                  3           2        1  
2            1          4                  3           2        3  
3            1          4                  2           3        5  
4            1          0                  3           0        1  

[5 rows x 23 columns]

In [33]:
df.shape

(8124, 23)

In [34]:
df.population.value_counts()

4    4040
5    1712
3    1248
2     400
0     384
1     340
Name: population, dtype: int64

In [35]:
# Cek kolom null
# Memeriksa apakah ada nilai yang hilang
df.isnull().sum()

class                       0
cap-shape                   0
cap-surface                 0
cap-color                   0
bruises                     0
odor                        0
gill-attachment             0
gill-spacing                0
gill-size                   0
gill-color                  0
stalk-shape                 0
stalk-root                  0
stalk-surface-above-ring    0
stalk-surface-below-ring    0
stalk-color-above-ring      0
stalk-color-below-ring      0
veil-type                   0
veil-color                  0
ring-number                 0
ring-type                   0
spore-print-color           0
population                  0
habitat                     0
dtype: int64

### Split data training dan testing

In [36]:
#Separating the independent varaibles
x = df.drop('population',axis=1)
x.head()

class  cap-shape  cap-surface  cap-color  bruises  odor  gill-attachment  \
0      1          5            2          4        1     6                1   
1      0          5            2          9        1     0                1   
2      0          0            2          8        1     3                1   
3      1          5            3          8        1     6                1   
4      0          5            2          3        0     5                1   

   gill-spacing  gill-size  gill-color  ...  stalk-surface-above-ring  \
0             0          1           4  ...                         2   
1             0          0           4  ...                         2   
2             0          0           5  ...                         2   
3             0          1           5  ...                         2   
4             1          0           4  ...                         2   

   stalk-surface-below-ring  stalk-color-above-ring  stalk-color-below-ring  \
0                         2                       7                       7   
1                         2                       7                       7   
2                         2                       7                       7   
3                         2                       7                       7   
4                         2                       7                       7   

   veil-type  veil-color  ring-number  ring-type  spore-print-color  habitat  
0          0           2            1          4                  2        5  
1          0           2            1          4                  3        1  
2          0           2            1          4                  3        3  
3          0           2            1          4                  2        5  
4          0           2            1          0                  3        1  

[5 rows x 22 columns]

In [37]:
#Separating the population varaible
y = df[['population']]
y.head()

population
0           3
1           2
2           2
3           3
4           0

In [38]:
from sklearn.model_selection import train_test_split

#Separating the dataset into training and testing dataset
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size = 0.3, random_state = 4)

In [39]:
y_train = np.array(y_train['population'])
y_test1 = np.array(y_test['population'])

### Traning Decision Tree

#### 1) Without Hyper parameter Optimization

In [40]:
#Fitting a decision tree with default hyper parameters
tree = DecisionTreeClassifier()
tree.fit(x_train,y_train)
pred_tree = tree.predict(x_test)

In [41]:
#Checking different metrics for decision tree model with default hyper parameters
print('Checking different metrics for decision tree model with default hyper parameters:\n')
print("Training accuracy: ",tree.score(x_train,y_train))
acc_score = accuracy_score(y_test, pred_tree)
print('Testing accuracy: ',acc_score)

Checking different metrics for decision tree model with default hyper parameters:

Training accuracy:  0.7460429124164615
Testing accuracy:  0.38474159146841674


#### 2) With Hyper parameter Optimization

In [42]:
#Setting values for the parameters
#n_estimators = [100, 300, 500, 800, 1200]
max_depth = [5, 10, 15, 25, 30]
min_samples_split = [2, 5, 10, 15, 100]
min_samples_leaf = [1, 2, 5, 10]
max_features = [1, 2, 5, 10]

#Creating a dictionary for the hyper parameters
hyperT = dict(max_depth = max_depth, min_samples_split = min_samples_split, 
              min_samples_leaf = min_samples_leaf, max_features=max_features)

#Applying GridSearchCV to get the best value for hyperparameters
gridT = GridSearchCV(tree, hyperT, cv = 3, verbose = 1, n_jobs = -1)
bestT = gridT.fit(x_train, y_train)

Fitting 3 folds for each of 400 candidates, totalling 1200 fits


In [43]:
#Printing the best hyperparameters
print('The best hyper parameters are: \n',gridT.best_params_)

The best hyper parameters are: 
 {'max_depth': 5, 'max_features': 5, 'min_samples_leaf': 1, 'min_samples_split': 100}


#### 3) Fitting Descision Tree with best Hyper parameters

In [44]:
#Fitting the decision tree model with the best hyper parameters obtained through GridSearchCV
tree1 = DecisionTreeClassifier(max_depth=5, min_samples_leaf=2, min_samples_split=5, max_features=10)
tree1.fit(x_train,y_train)
pred_tree1 = tree1.predict(x_test)

In [45]:
#Checking different metrics for decision tree model after tuning the hyperparameters
print('Checking different metrics for decision tree model after tuning the hyperparameters:\n')
print("Training accuracy: ",tree1.score(x_train,y_train))
acc_score = accuracy_score(y_test, pred_tree1)
print('Testing accuracy: ',acc_score)

Checking different metrics for decision tree model after tuning the hyperparameters:

Training accuracy:  0.6348927189588462
Testing accuracy:  0.6341263330598852


### Training RandomForest

#### 1) Without Hyper parameter Optimization

In [46]:
#Fitting a RandomForest with default hyper parameters
rf = RandomForestClassifier()
rf.fit(x_train,y_train)
pred_rf = rf.predict(x_test)

In [47]:
#Checking different metrics for decision random forest model with default hyper parameters
print('Checking different metrics for decision random forest model with default hyper parameters:\n')
print("Training accuracy: ",rf.score(x_train,y_train))
acc_score = accuracy_score(y_test, pred_rf)
print('Testing accuracy: ',acc_score)

Checking different metrics for decision random forest model with default hyper parameters:

Training accuracy:  0.7460429124164615
Testing accuracy:  0.38474159146841674


#### 2) With Hyper parameter Optimization

In [48]:
#Setting values for the parameters
#n_estimators = [100, 300, 500, 800, 1200]
max_depth = [5, 10, 15, 25, 30]
min_samples_split = [2, 5, 10, 15, 100]
min_samples_leaf = [1, 2, 5, 10]
max_features = [1, 2, 5, 10]

#Creating a dictionary for the hyper parameters
hyper_rf = dict(max_depth = max_depth, min_samples_split = min_samples_split, min_samples_leaf = min_samples_leaf, max_features=max_features)

#Applying GridSearchCV to get the best value for hyperparameters
gridrf = GridSearchCV(rf, hyper_rf, cv = 3, verbose = 1, n_jobs = -1)
bestrf = gridrf.fit(x_train, y_train)

Fitting 3 folds for each of 400 candidates, totalling 1200 fits


In [49]:
#Printing the best hyperparameters
print('The best hyper parameters are: \n',gridrf.best_params_)

The best hyper parameters are: 
 {'max_depth': 5, 'max_features': 1, 'min_samples_leaf': 2, 'min_samples_split': 5}


#### 3) Fitting RandomForest with best Hyper parameters

In [50]:
#Fitting the decision tree model with the best hyper parameters obtained through GridSearchCV
rf2 = DecisionTreeClassifier(max_depth=5, min_samples_leaf=2,min_samples_split=5, max_features=1)
rf2.fit(x_train,y_train)
pred_tree1 = rf2.predict(x_test)

In [51]:
#Checking different metrics for decision tree model after tuning the hyperparameters
print('Checking different metrics for decision tree model after tuning the hyperparameters:\n')
print("Training accuracy: ",rf2.score(x_train,y_train))
acc_score = accuracy_score(y_test, pred_tree1)
print('Testing accuracy: ',acc_score)

Checking different metrics for decision tree model after tuning the hyperparameters:

Training accuracy:  0.6204713330988393
Testing accuracy:  0.6218211648892535


### Training AdaBoost

#### 1) Without Hyper parameter Optimization

In [52]:
abc = AdaBoostClassifier()
abc.fit(x_train,y_train)
pred_abc = abc.predict(x_test)

In [53]:
#Checking different metrics for AdaBoost model with default hyper parameters
print('Checking different metrics for AdaBoost model with default hyper parameters:\n')
print("Training accuracy: ",abc.score(x_train,y_train))
acc_score = accuracy_score(y_test, pred_abc)
print('Testing accuracy: ',acc_score)

Checking different metrics for AdaBoost model with default hyper parameters:

Training accuracy:  0.3855082659162856
Testing accuracy:  0.3921246923707957


#### 2) With Hyper parameter Optimization

In [54]:
#Setting values for the parameters
n_estimators = [50, 100, 150, 200]

#Creating a dictionary for the hyper parameters
hyper_abc = dict(n_estimators=n_estimators)

#Applying GridSearchCV to get the best value for hyperparameters
gridabc = GridSearchCV(abc, hyper_abc, cv = 3, verbose = 1, n_jobs = -1)
bestabc = gridabc.fit(x_train, y_train)

Fitting 3 folds for each of 4 candidates, totalling 12 fits


In [55]:
#Printing the best hyperparameters
print('The best hyper parameters are: \n',gridabc.best_params_)

The best hyper parameters are: 
 {'n_estimators': 100}


#### 3) Fitting AdaBoost with best Hyper parameters

In [56]:
#Fitting the AdaBoost model with the best hyper parameters obtained through GridSearchCV
abc2 = AdaBoostClassifier(n_estimators=100)
abc2.fit(x_train,y_train)
pred_abc1 = abc2.predict(x_test)

In [57]:
#Checking different metrics for AdaBoost model after tuning the hyperparameters
print('Checking different metrics for AdaBoost model after tuning the hyperparameters:\n')
print("Training accuracy: ",abc2.score(x_train,y_train))
acc_score = accuracy_score(y_test, pred_abc1)
print('Testing accuracy: ',acc_score)

Checking different metrics for AdaBoost model after tuning the hyperparameters:

Training accuracy:  0.3855082659162856
Testing accuracy:  0.3921246923707957


### Evaluasi

#### Decision Tree.
- sebelum hyperparameters
  - Kita dapat melihat bahwa akurasi pelatihan Decision Tree adalah 0.7460429124164615 (74%). Saat menyesuaikan model yang sama untuk menguji data, akurasi turun menjadi sekitar 0.38474159146841674 (38%). Model tersebut memiliki data pelatihan yang berlebihan, dan variansnya sangat tinggi, yang menunjukkan bahwa model berperilaku berbeda untuk sampel yang berbeda.  
- sesudah hyperparameters
  - Ketika menerapkan algoritma GridSearchCV ke model yang digunakan untuk memilih hyperparameter terbaik, terlihat bahwa akurasi pelatihan sekitar 0.6315511783327471 (63%) dan akurasi pengujian sekitar 0.6222313371616078 (62%). Sementara penyetelan hyperparameters mengurangi akurasi pelatihan, itu juga mengurangi overfitting di Decision Tree.

#### Random Forest.
- sebelum hyperparameters
  - Kita dapat melihat bahwa akurasi pelatihan Random Forest adalah 0.7460429124164615 (74%). Saat menyesuaikan model yang sama untuk menguji data, akurasi turun menjadi sekitar 0.38474159146841674 (38%). Model tersebut memiliki data pelatihan yang berlebihan, dan variansnya sangat tinggi, yang menunjukkan bahwa model berperilaku berbeda untuk sampel yang berbeda.  
- sesudah hyperparameters
  - Ketika menerapkan algoritma GridSearchCV ke model yang digunakan untuk memilih hyperparameter terbaik, terlihat bahwa akurasi pelatihan sekitar 0.6204713330988393 (62%) dan akurasi pengujian sekitar 0.6218211648892535 (62%). Sementara penyetelan hyperparameters mengurangi akurasi pelatihan, itu juga mengurangi overfitting di Random Forest.

#### AdaBoost.
- sebelum hyperparameters
  - Kita dapat melihat bahwa akurasi pelatihan Random Forest adalah 0.3855082659162856 (38%). Saat menyesuaikan model yang sama untuk menguji data, akurasi turun menjadi sekitar 0.3921246923707957 (39%). Model tersebut memiliki data pelatihan yang berlebihan, dan variansnya sangat tinggi, yang menunjukkan bahwa model berperilaku berbeda untuk sampel yang berbeda.  
- sesudah hyperparameters
  - Ketika menerapkan algoritma GridSearchCV ke model yang digunakan untuk memilih hyperparameter terbaik, terlihat bahwa akurasi pelatihan sekitar 0.3855082659162856 (38%) dan akurasi pengujian sekitar 0.3921246923707957 (39%). Sementara penyetelan hyperparameters mengurangi akurasi pelatihan, itu juga mengurangi overfitting di Random Forest.